api/types/appserver.go

```diff
@@ -295,16 +296,90 @@ func (s *AppServerV3) MatchSearch(values []string) bool {
 	return MatchSearch(nil, values, nil)
 }
 
-// AppServers represents a list of app servers.
-type AppServers []AppServer
+type AppServerSorter struct {
+	servers []AppServer
+	lessFn  func(i, j int) bool
+}
+
+// AppServers returns a sorter that implements the Sort interface,
+// Call its Sort method to sort the data by sort criteria.
+func AppServers(servers []AppServer) *AppServerSorter {
+	return &AppServerSorter{
+		servers: servers,
+	}
+}
+
+// Len is part of sort.Interface.
+func (s *AppServerSorter) Len() int { return len(s.servers) }
+
+// Less is part of sort.Interface.
+func (s *AppServerSorter) Less(i, j int) bool { return s.lessFn(i, j) }
+
+// Swap is part of sort.Interface.
+func (s *AppServerSorter) Swap(i, j int) { s.servers[i], s.servers[j] = s.servers[j], s.servers[i] }
 
-// Len returns the slice length.
-func (s AppServers) Len() int { return len(s) }
+// Sort sorts a list of app servers according to the sort criteria.
+func (s *AppServerSorter) Sort(sortBy *SortBy) error {
+	if sortBy == nil {
+		return nil
+	}
+
+	// We assume sorting by type AppServer, we are really
+	// wanting to sort its contained resource Application.
+	switch sortBy.Field {
+	case ResourceMetadataName:
+		s.lessFn = func(i, j int) bool {
+			return compareStrByDir(s.servers[i].GetApp().GetName(), s.servers[j].GetApp().GetName(), sortBy.Dir)
+		}
+	case ResourceSpecDescription:
+		s.lessFn = func(i, j int) bool {
+			return compareStrByDir(s.servers[i].GetApp().GetDescription(), s.servers[j].GetApp().GetDescription(), sortBy.Dir)
+		}
+	case ResourceSpecPublicAddr:
+		s.lessFn = func(i, j int) bool {
+			return compareStrByDir(s.servers[i].GetApp().GetPublicAddr(), s.servers[j].GetApp().GetPublicAddr(), sortBy.Dir)
+		}
+	default:
+		return trace.NotImplemented("sorting by field %q for resource %q is not supported", sortBy.Field, KindAppServer)
+	}
 
-// Less compares app servers by name and host ID.
-func (s AppServers) Less(i, j int) bool {
-	return s[i].GetName() < s[j].GetName() && s[i].GetHostID() < s[j].GetHostID()
+	sort.Sort(s)
+	return nil
+}
+
+// AsResources returns app servers as type resources with labels.
+func (s *AppServerSorter) AsResources() []ResourceWithLabels {
+	resources := make([]ResourceWithLabels, len(s.servers))
+	for i, server := range s.servers {
+		resources[i] = ResourceWithLabels(server)
+	}
```

Comment:
@espadolini, I am curious why we prefer the `append` way? I thought if we knew the length does not change, assignment by index is faster than appending?

Reply:
It's a good habit to get into; it doesn't matter here, but if the body of the loop is more complicated with early returns or other weird flow control, appending to a slice with capacity but zero length makes it so that the elements in the slice are always fully initialized and valid.